### PR TITLE
Fix issues with extra storage permissions.

### PR DESCRIPTION
**Pull request by ewelinajozwik**
Fixes [FNS-75](https://thehyve.atlassian.net/browse/FNS-75)

[FNS-75]: https://thehyve.atlassian.net/browse/FNS-75?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: projects/saturn/src/main/java/io/fairspace/saturn/webdav/DavFactory.java [hunk 1]
```diff
@@ -60,6 +60,9 @@ public Resource getResource(org.apache.jena.rdf.model.Resource subject) {
     }
 
     public Resource getResource(org.apache.jena.rdf.model.Resource subject, Access access) {
+        if (isExtraStoreResource() && subject.hasProperty(RDF.type, FS.File) && !subject.hasProperty(FS.createdBy, currentUserResource())) {
```
Comment (frankyhollywood) [inline projects/saturn/src/main/java/io/fairspace/saturn/webdav/DavFactory.java:63]: Having 'subject.hasProperty(RDF.type, FS.File)' means that a non-file resource (directory, collection) can be accessed by everyone.

File: projects/saturn/src/main/java/io/fairspace/saturn/webdav/DavFactory.java [hunk 2]
```diff
@@ -136,18 +143,19 @@ public Access getAccess(org.apache.jena.rdf.model.Resource subject) {
...
             return Access.List;
         }
 
         return access;
     }
 
     private Access getExtraStorageAccess(org.apache.jena.rdf.model.Resource subject) {
-        if (subject.hasProperty(FS.createdBy, currentUserResource()) && subject.hasProperty(RDF.type, FS.File)) {
-            return Access.Manage;
+        if (subject.equals(this.rootSubject)) {
```
Comment (frankyhollywood) [inline projects/saturn/src/main/java/io/fairspace/saturn/webdav/DavFactory.java:154]: could the order of if statements have a result that you get read access, while you should have write access?

File: projects/saturn/src/main/java/io/fairspace/saturn/webdav/resources/BaseResource.java [hunk 3]
```diff
@@ -94,8 +94,19 @@ public Date getCreateDate() {
...
+        boolean purge;
+        if (factory.isExtraStoreResource()) {
+            if (subject.hasProperty(RDF.type, FS.ExtraStorageDirectory)) {
+                throw new NotAuthorizedException("Not authorized to purge the extra store root directory.", this, SC_FORBIDDEN);
+            }
+            purge = true;
+        } else {
+            purge = subject.hasProperty(FS.dateDeleted);
+            if (purge && !factory.userService.currentUser().isAdmin()) {
+                throw new NotAuthorizedException("Not authorized to purge the resource.", this, SC_FORBIDDEN);
```
Comment (frankyhollywood) [inline projects/saturn/src/main/java/io/fairspace/saturn/webdav/resources/BaseResource.java:107]: why does the non-extraStorageResource path an extra validation + exception, was the old good not suffictient?

**Review by frankyhollywood — APPROVED** [review]

File: projects/saturn/src/main/java/io/fairspace/saturn/webdav/DavFactory.java (hunk 1 above)
Comment (ewelinajozwik) [reply projects/saturn/src/main/java/io/fairspace/saturn/webdav/DavFactory.java:63]: yes, I though for now there will be simple file-based permissions , as this is what we need for jupyter. In the future we can make extra storage permissions more complex (e.g. when we find another usecase for this storage).

File: projects/saturn/src/main/java/io/fairspace/saturn/webdav/DavFactory.java (hunk 2 above)
Comment (ewelinajozwik) [reply projects/saturn/src/main/java/io/fairspace/saturn/webdav/DavFactory.java:154]: yes, the order matters, this is to protect the root folder being deleted by a user (it is created on application startup and should not be deleted), so it has to be read-only.

File: projects/saturn/src/main/java/io/fairspace/saturn/webdav/resources/BaseResource.java (hunk 3 above)
Comment (ewelinajozwik) [reply projects/saturn/src/main/java/io/fairspace/saturn/webdav/resources/BaseResource.java:107]: the validation was there, but I just moved it from `protected void delete(boolean purge)` to `public final void delete()` function, so it is done earlier on and so I can overwrite the protected one in other classes without repeating the validation in all the places.

**Pull request merged**